cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -64,14 +64,26 @@ private predicate operandToInstructionTaintStep(Operand opFrom, Instruction inst
     or
     instrTo instanceof PointerArithmeticInstruction
     or
-    instrTo.(FieldAddressInstruction).getField().getDeclaringType() instanceof Union
-    or
     // The `CopyInstruction` case is also present in non-taint data flow, but
     // that uses `getDef` rather than `getAnyDef`. For taint, we want flow
     // from a definition of `myStruct` to a `myStruct.myField` expression.
     instrTo instanceof CopyInstruction
   )
   or
+  // Unary instructions tend to preserve enough information in practice that we
+  // want taint to flow through.
+  // The exception is `FieldAddressInstruction`. Together with the rules below for
+  // `LoadInstruction`s and `ChiInstruction`s, flow through `FieldAddressInstruction`
+  // could cause flow into one field to come out an unrelated field.
```

Comment:
I think it can happen even without a function call boundary:
```
r0 = ...
r1 = FieldAddress[f1] : r0
r2 = load : r1, ~m1
```
if we have flow from `r0` to `r1` then we'll have flow from `r0` to `r2` due to the `instrTo.(LoadInstruction).getSourceAddressOperand() = opFrom` rule.

Reply:
I think having flow from `r0` to `r2` is OK by itself - If `r0` is a pointer that's computed using tainted data, then `r2` is a load from that tainted pointer and therefore also tainted. It's only when `r0` is tainted because of a pointer/object conflation that having flow from `r0` to `r2` is problematic.